src/reusedBrowser.ts

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ChildProcess, spawn } from 'child_process';
+import { PlaywrightTest, TestConfig } from './playwrightTest';
+import { SidebarViewProvider } from './sidebarView';
+import { createGuid } from './utils';
+import * as vscodeTypes from './vscodeTypes';
+
+export class ReusedBrowser implements vscodeTypes.Disposable {
+  private _vscode: vscodeTypes.VSCode;
+  private _sidebarView!: SidebarViewProvider;
+  private _playwrightTest: PlaywrightTest;
+  private _serverProcess: ChildProcess | undefined;
+  private _selectorExplorerBox: vscodeTypes.InputBox | undefined;
+
+  constructor(vscode: vscodeTypes.VSCode, context: vscodeTypes.ExtensionContext, sidebarView: SidebarViewProvider, playwrightTest: PlaywrightTest) {
+    this._vscode = vscode;
+    this._sidebarView = sidebarView;
+    this._playwrightTest = playwrightTest;
+    const disposables = [
+      this._sidebarView.onDidChangeReuseBrowser(async reuseBrowser => {
+        if (!reuseBrowser)
+          this.stop();
+      }),
+      this._sidebarView.onDidRequestInspect(() => this.inspect()),
+      this,
+    ];
+    context.subscriptions.push(...disposables);
+  }
+
+  dispose() {
+    this.stop();
+  }
+
+  async startIfNeeded(config: TestConfig) {
+    if (this._serverProcess || !this._sidebarView.reuseBrowser())
+      return;
+
+    if (config.version < 1.25) {
+      this._vscode.window.showErrorMessage(`Playwright v1.25+ is required for browser reuse, v${config.version} found`);
+      return;
+    }
+    const node = await this._playwrightTest.findNode();
+    const allArgs = [
+      config.cli,
+      'run-server',
+      '--reuse-browser',
+      `--path=/${createGuid()}`
+    ];
+    this._serverProcess = spawn(node, allArgs, {
+      cwd: config.workspaceFolder,
+      stdio: ['pipe', 'pipe', 'pipe', 'ipc'],
+      env: { ...process.env, PW_CODEGEN_NO_INSPECTOR: '1' },
+    });
+
+    this._serverProcess.on('exit', () => {
+      this._setBrowserServerWS(undefined);
+      this._serverProcess = undefined;
```

Comment:
We should resolve/reject the promise from line 79 here to avoid stalling.

Reply:
Sure